### PR TITLE
Dev add evaluation

### PR DESCRIPTION
**Pull request by JoshuaGRubin**
- Added a base class to derive models from
- Derived a baseline model that characterizes users by the normalized mean embedding of a corpus of tweets.
- Added a generalized evaluation module.  Includes:
  - loading data
  - shuffling
  - doing train/test split
  - computing confusion matrix by user.

### CODE REVIEW

File: source/models/averaged_cosine_sim_model.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class AvgCosSimModel(Model):
+    """ Derives a model that characterizes the corpus with a single normalized
+        mean vector at initialization.  Inference is performed by cosine
+        similarity with a threshold to produce true for inconsistent/fraudulent
+        and false for consistent/authentic.
+    """
+    def __init__(self, corpus=None, embedded_corpus=None):
+        """ Requires a pre-embedded corpus passed in through <embedded_corpus>.
+        """
```
Comment (mstefferson) [inline source/models/averaged_cosine_sim_model.py:25]: Doc strings should include input arguments. E.g.
```
"""Requires a pre-embedded corpus passed in through <embedded_corpus>.

Args:

corpus (str): corpus identifier
embedded_corpus (np.array): embedding vector for corpus
"""
```

File: source/models/averaged_cosine_sim_model.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        if embedded_corpus is None:
```
Comment (mstefferson) [inline source/models/averaged_cosine_sim_model.py]: Why default it to None then? Can't you just make it a positional argument?

File: source/models/averaged_cosine_sim_model.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            raise Exception('Sorry, this model requires an embedded_corpus.')
+            
+        mean_of_embeddeded_vecs = (np.asarray([x for x in embedded_corpus])
+                                    .mean(axis=0))    
+        self.user_avg_embedding = mean_of_embeddeded_vecs/np.sqrt(
+                                          np.inner(mean_of_embeddeded_vecs,
+                                                   mean_of_embeddeded_vecs))
+        # Default setting; change with set_hyperparameter on base-class
+        self.params = {'threshold':0.3}
```
Comment (mstefferson) [inline source/models/averaged_cosine_sim_model.py:33]: Should you pass in threshold as a parameter. Currently default it to 0.3

File: source/models/averaged_cosine_sim_model.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            and input embedded tweet.  Must (for now) pass a pre-embedded
+            list of tweets to <embedded_tweets>.
+            
+            Return value:
+            Array of true/false values corresponding to fraud/authentic.
+        """
+        
+        return [ np.inner(self.user_avg_embedding,
+                          np.asarray(v)) < self.params['threshold']
+                          for v in embedded_tweets            ]
```
Comment (mstefferson) [inline source/models/averaged_cosine_sim_model.py]: There is something weird going on with your end-of-line 
image

File: source/models/base_model.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        pass
+   
+    def set_hyperparameters(self, params = {}):
+        """ Set model hyperparameters e.g. classifier threshold."""
+        self.params = params
+    
+    def infer(self, tweets=None, embedded_tweets=None):
+        """ Pass in a list of tweets (text or embedded if appropriate for model
+            type) to test against 
+        """
```
Comment (mstefferson) [inline source/models/base_model.py]: Is this suppose to be blank?

File: source/models/base_model.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+embedding-bassed or otherwise (e.g. TFIDF).
+
+June, 2019
+@author: Joshua Rubin
+"""
+
+class Model:
+    """ Base class for binary classifier to identify fraudulent tweet . """
+   
+    def __init__(self, corpus=None, embedded_corpus=None):
```
Comment (mstefferson) [inline source/models/base_model.py]: Why have an `__init__` here is it just passes?

File: source/models/evaluate_model.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        if file[0] == '@':
+            print(file)
+            newFrame = pd.read_json(os.path.join(directory_path,file))
+            newFrame.columns = ['tweet','date','embedding']
+            newFrame['name'] = file.split('.')[0]
+            frames.append(newFrame)
+        
+    allData = pd.concat(frames)
+
+    return np.split(allData.sample(frac=1),
```
Comment (mstefferson) [inline source/models/evaluate_model.py]:
```
from sklearn.model_selection import train_test_split

train, test = train_test_split(df, test_size=0.2)
```

Comment (mstefferson) [reply source/models/evaluate_model.py]: Don't need to write everything yourself. It's also easier for other people because they don't have to wonder if your homemade split works correctly

File: source/models/evaluate_model.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+# Helpers to compute true and false positive rates form confusion matrix
+def TPR(c): return c[0,0]/(c[0,0]+c[0,1])
+def FPR(c): return c[1,0]/(c[1,0]+c[1,1])
+
+# If I'm being run as a script:
+if __name__ == '__main__':
+    
+    with open(CONFIG_PATH, 'r') as file:
+        config =  json.loads(file.read())
```
Comment (mstefferson) [inline source/models/evaluate_model.py:81]: config. Nice!

File: source/models/evaluate_model.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        print('%8s %6s %6s' % ('Thresh', 'TPR', 'FPR'))
+
+        for model_thresh in np.arange(0.2,0.55,0.05):
+    
+            # Add/update a column in test_data called 'is_fraud' that is false
+            # if the tweet was from <user> and true otherwise.
+            test_data['is_fraud'] = (test_data['name'] != user)
+        
+            train_user_embs = train_data[train_data['name'] == user
+                                                             ]['embedding']
```
Comment (mstefferson) [inline source/models/evaluate_model.py]: ` ]['embedding']` a typo?

Comment (mstefferson) [reply source/models/evaluate_model.py]: Nevermind. Just a little hard to read

File: source/models/evaluate_model.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+# Helpers to compute true and false positive rates form confusion matrix
+def TPR(c): return c[0,0]/(c[0,0]+c[0,1])
+def FPR(c): return c[1,0]/(c[1,0]+c[1,1])
+
+# If I'm being run as a script:
+if __name__ == '__main__':
+    
+    with open(CONFIG_PATH, 'r') as file:
+        config =  json.loads(file.read())
+
```
Comment (mstefferson) [inline source/models/evaluate_model.py:82]: It would be nice if `if __name__ == '__main__'` called a function containing the code below here. I.e., put this in a new function (called `main` or something). This will allow it to be called through over means if it needs to be.

**Review by mstefferson — COMMENTED** [review]
Glad to see the evaluation in place! A few comments, but keep on moving!

File: source/models/averaged_cosine_sim_model.py (hunk 1 above)
Comment (JoshuaGRubin) [reply source/models/averaged_cosine_sim_model.py:25]: Sure.  Fixing.

File: source/models/averaged_cosine_sim_model.py (hunk 2 above)
Comment (JoshuaGRubin) [reply source/models/averaged_cosine_sim_model.py]: Good question, @mstefferson.  There might be a better way to do this, but I'm trying to write an interface that accommodates scenarios where a model could take a raw string (e.g. TDIDF) instead of (or alternatively to) a embedded vector input.  I'd like for all models to have the same method prototypes.

File: source/models/averaged_cosine_sim_model.py (hunk 3 above)
Comment (JoshuaGRubin) [reply source/models/averaged_cosine_sim_model.py:33]: This is modified though the base_class's `set_hyperparameter`.  It can be changed without creating a new object (see evaluate_model.py).

File: source/models/base_model.py (hunk 5 above)
Comment (JoshuaGRubin) [reply source/models/base_model.py]: Will fix.  Not sure how that didn't throw an error.

File: source/models/base_model.py (hunk 6 above)
Comment (JoshuaGRubin) [reply source/models/base_model.py]: Just a loose end - will take out.  Was trying to define a standard interface for models... but not 100% sure how best to do that with Python classes.

File: source/models/averaged_cosine_sim_model.py (hunk 4 above)
Comment (JoshuaGRubin) [reply source/models/averaged_cosine_sim_model.py]: I'm guessing that this means that I don't have a new-line character at the end of file; apparently nonstandard.  Fixing.

File: source/models/averaged_cosine_sim_model.py (hunk 2 above)
Comment (JoshuaGRubin) [reply source/models/averaged_cosine_sim_model.py]: Simplifying as suggested (rather than outthinking a problem I don't have yet).

File: source/models/averaged_cosine_sim_model.py (hunk 3 above)
Comment (JoshuaGRubin) [reply source/models/averaged_cosine_sim_model.py:33]: Great suggestion.  Definitely cleaner-looking your way.
